### PR TITLE
Re-implements the Bubble widget.

### PR DESCRIPTION
**Pull request by AnthonyZi**
This commit comprises three commits:
- Re-implements the Bubble widge
- Implements tests for the Bubble widget
- Adapts existing code to Bubble changes

The old Bubble widget design was complicated and used a GridLayout with
multiple dummy widgets solely to get an arrow placed as desired.
Multiple magic numbers like 99 columns or 99 rows were undocumented
and unintuitive. Probably some of these numbers even add unnecessary
limitations to the Bubble widget.
The Bubble widget was fused with the 'BubbleContent' which by design
makes the Bubble widget feel very unflexible.
Sizing the Bubble according to the content was unintuitive if not even
unpossible for some cases and the arrow positioning was relying on magic
numbers as well. Lastly, the old implementation lacked test coverage
with very few tests that did not even cover the core logic of the
widget.

The new implementation reduces the core logic (formerly "on_arrow_pos",
~100 lines of code) to ~30 lines plus a well documented lookup structure
without magic but intuitive values. Furthermore, it adds proper inline
documentation for maintainers.

This PR also adds unit tests for the core logic of the Bubble widget.

Maintainer merge checklist
* [x] Title is descriptive/clear for inclusion in release notes.
* [x] Applied a `Component: xxx` label.
* [x] Applied the `api-deprecation` or `api-break` label.
* [x] Applied the `release-highlight` label to be highlighted in release notes.
* [x] Added to the milestone version it was merged into.
* [x] **Unittests** are included in PR.
* [x] Properly documented, including `versionadded`, `versionchanged` as needed.

### CODE REVIEW

File: kivy/uix/bubble.py [hunk 1]
```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
...
+        elif base_layout_key == "right_mid":
+            arrow_layout[-1] = {"left": 0.0, "y": calc_x0(y, self.height)}
+        return arrow_layout
+
+    def update_content_size(self, instance, value):
+        self.content_size = self.content.size
+
+    def adjust_position(self):
+        if self.limit_to is not None and not self._temporarily_ignore_limits:
+            if self.limit_to is EventLoop.window:
```
Comment (AnthonyZi) [inline kivy/uix/bubble.py:418]: This part was mainly copied from the old implementation. I use the Bubble widget mainly in a Scatter Plane where this logic does somehow not really make sense. Maybe someone else can verify that this code still "usefull" in other use cases.

File: kivy/uix/bubble.py [hunk 2]
```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
...
+                lim_top = self.limit_to.top
+                lim_right = self.limit_to.right
+
+            self._temporarily_ignore_limits = True
 
-        # add widgets to arrow_layout
-        add = self_arrow_layout.add_widget
-        for widg in arrow_list:
-            add(widg)
+            if not (lim_x > self.x and lim_right < self.right):
```
Comment (AnthonyZi) [inline kivy/uix/bubble.py:429]: Adapted this part a little to make it only "intelligent" if there is a possible better solution (widget is completely within the limits).

**Comment by AnthonyZi** [comment]
Hey, this is my first PR here. I am wondering if there is something blocking the review process. Is someone/something waiting for me to take some action? Tanks in advance!

File: kivy/uix/bubble.py [hunk 3]
```diff
@@ -25,82 +39,67 @@
...
 
-class BubbleContent(GridLayout):
-    pass
-
-
-class Bubble(GridLayout):
-    '''Bubble class. See module documentation for more information.
-    '''
-
+class BubbleContent(BoxLayout):
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:135]: I updated the documentation. Updates include this request.

File: kivy/uix/bubble.py [hunk 4]
```diff
@@ -115,22 +114,54 @@ class Bubble(GridLayout):
...
+        'atlas://data/images/defaulttheme/bubble_arrow'
+    )
     ''' Image of the arrow pointing to the bubble.
 
     :attr:`arrow_image` is a :class:`~kivy.properties.StringProperty` and
     defaults to 'atlas://data/images/defaulttheme/bubble_arrow'.
     '''
 
+    arrow_color = ColorProperty([1, 1, 1, 1])
+    '''Arrow color, in the format (r, g, b, a). To use it you have to set
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:222]: I updated the documentation. Updates include this request.

File: kivy/uix/bubble.py [hunk 5]
```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
...
+    '''Specifies the position of the arrow as predefined relative position to
+    the bubble.
     Can be one of: left_top, left_mid, left_bottom top_left, top_mid, top_right
     right_top, right_mid, right_bottom bottom_left, bottom_mid, bottom_right.
 
     :attr:`arrow_pos` is a :class:`~kivy.properties.OptionProperty` and
     defaults to 'bottom_mid'.
     '''
+    flex_arrow_pos = ListProperty(None)
+    '''Specifies the position of the arrow as flex coordinate around the
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:259]: I updated the documentation. Updates include this request.

File: kivy/uix/bubble.py [hunk 6]
```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
...
+    '''The width of the content Widget.
+
+    :attr:`content_width` is a :class:`~kivy.properties.NumericProperty` and
+    is the same as self.content.width if content is not None, else it defaults
+    to 0. It is read only.
+    '''
+
+    content_height = NumericProperty(0)
+    '''The height of the content Widget.
+
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:335]: I updated the documentation. Updates include this request.

File: kivy/uix/bubble.py [hunk 7]
```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
...
+    :attr:`content_height` is a :class:`~kivy.properties.NumericProperty` and
+    is the same as self.content.height if content is not None, else it defaults
+    to 0. It is read only.
+    '''
+
+    content_size = ReferenceListProperty(content_width, content_height)
+    ''' The size of the content Widget.
+
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:345]: I updated the documentation. Updates include this request.

File: kivy/uix/bubble.py [hunk 8]
```diff
@@ -25,82 +39,67 @@
...
     background_color = ColorProperty([1, 1, 1, 1])
     '''Background color, in the format (r, g, b, a). To use it you have to set
-    either :attr:`background_image` or :attr:`arrow_image` first.
+    :attr:`background_image` first.
 
     :attr:`background_color` is a :class:`~kivy.properties.ColorProperty` and
     defaults to [1, 1, 1, 1].
+    '''
 
-    .. versionchanged:: 2.0.0
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:101]: The properties 'background_image', 'background_color', 'border' and 'border_auto_scale' have been removed from the 'Bubble'. They were added as new properties to the widget 'BubbleContent'. Therefore, I would argue to remove 'versionchanged' documentation and add 'versionadded' where the properties have been added.
I changed that with my last commit, please check if that is ok.

File: kivy/uix/bubble.py [hunk 9]
```diff
@@ -140,263 +171,322 @@ class Bubble(GridLayout):
...
-    :class:`~kivy.properties.OptionProperty` and defaults to
-    'both_lower'.
+    :attr:`arrow_margin` is a :class:`~kivy.properties.ReferenceListProperty`
+    of (:attr:`arrow_margin_x`, :attr:`arrow_margin_y`) properties.
+    It is read only.
+    '''
+
+    content_width = NumericProperty(0)
+    '''The width of the content Widget.
+
```
[the remark this answers is not in the document]
Comment (AnthonyZi) [reply kivy/uix/bubble.py:325]: I updated the documentation. Updates include this request.

**Review by akshayaurora — APPROVED** [review]

**Comment by akshayaurora** [comment]
@AnthonyZi thanks 👍

**Pull request merged**